File: ureka-framework-main/ureka_framework/model/data_model/current_session.py
# Data Model (RAM)
from typing import Optional
from pydantic import BaseModel, ConfigDict, ValidationError

# Resource (Logger)
from ureka_framework.resource.logger.simple_logger import simple_log


######################################################
# Data Model
######################################################
class CurrentSession(BaseModel):
    # Access UT
    current_u_ticket_id: Optional[str] = None
    current_device_id: Optional[str] = None
    current_holder_id: Optional[str] = None
    current_task_scope: Optional[str] = None

    # CR-KE
    challenge_1: Optional[str] = None
    challenge_2: Optional[str] = None
    key_exchange_salt_1: Optional[str] = None
    key_exchange_salt_2: Optional[str] = None

    # PS
    current_session_key_str: Optional[str] = None

    plaintext_cmd: Optional[str] = None
    associated_plaintext_cmd: Optional[str] = None
    iv_cmd: Optional[str] = None
    ciphertext_cmd: Optional[str] = None
    gcm_authentication_tag_cmd: Optional[str] = None

    plaintext_data: Optional[str] = None
    associated_plaintext_data: Optional[str] = None
    iv_data: Optional[str] = None
    ciphertext_data: Optional[str] = None
    gcm_authentication_tag_data: Optional[str] = None

    # By default, Pydantic "ignore" extra input fields not defined in model schema
    # Moreover, we can explicitly "allow" or "forbid (with Error)" extra input fields not defined in model schema
    model_config = ConfigDict(extra="forbid")


################################################################################
#                             < CurrentSession_obj >                           #
#                                       | self-defined serialization           #
#                                       | (all str, which is native type)      #
#                                       v                                      #
#         < JSON_dict (Should be JSON serializable, i.e. native type) >        #
#                                       |                                      #
#                                       v                                      #
#                       < JSON_str (Printable Characters) >                    #
################################################################################
def current_session_to_jsonstr(current_session_obj: CurrentSession) -> str:
    # "indent" do not affect json validation, but may affect json size!?
    current_session_json = current_session_obj.model_dump_json(
        indent=4, exclude_none=True
    )
    return current_session_json


def jsonstr_to_current_session(json_str: str) -> CurrentSession:
    try:
        return CurrentSession.model_validate_json(json_str)
    except ValidationError as error:
        failure_msg = "NOT VALID JSON or VALID SCHEMA"
        # simple_log("error", f"{failure_msg}: {error}")
        raise RuntimeError(failure_msg)

File: ureka-framework-java/app/src/main/java/ureka/framework/model/data_model/CurrentSession.java
package ureka.framework.model.data_model;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.util.Map;

public class CurrentSession {
    private static final Gson gson = new GsonBuilder().setPrettyPrinting().serializeNulls().create();
    // Access UT
    private String currentUTicketId = null;
    private String currentDeviceId = null;
    private String currentHolderId = null;
    private String currentTaskScope = null;
    // CR-KE
    private String challenge1 = null;
    private String challenge2 = null;
    private String keyExchangeSalt1 = null;
    private String keyExchangeSalt2 = null;
    // PS
    private String currentSessionKeyStr = null;

    private String plaintextCmd = null;
    private String associatedPlaintextCmd = null;
    private String ivCmd = null;
    private String ciphertextCmd = null;
    private String gcmAuthenticationTagCmd = null;

    private String plaintextData = null;
    private String associatedPlaintextData = null;
    private String ivData = null;
    private String ciphertextData = null;
    private String gcmAuthenticationTagData = null;

    public CurrentSession() {}
    public CurrentSession(Map<String, String> values) {
        this.currentUTicketId = values.getOrDefault("currentUTicketId", null);
        this.currentDeviceId = values.getOrDefault("currentDeviceId", null);
        this.currentHolderId = values.getOrDefault("currentHolderId", null);
        this.currentTaskScope = values.getOrDefault("currentTaskScope", null);
        this.challenge1 = values.getOrDefault("challenge1", null);
        this.challenge2 = values.getOrDefault("challenge2", null);
        this.keyExchangeSalt1 = values.getOrDefault("keyExchangeSalt1", null);
        this.keyExchangeSalt2 = values.getOrDefault("keyExchangeSalt2", null);
        this.currentSessionKeyStr = values.getOrDefault("currentSessionKeyStr", null);
        this.plaintextCmd = values.getOrDefault("plaintextCmd", null);
        this.associatedPlaintextCmd = values.getOrDefault("associatedPlaintextCmd", null);
        this.ciphertextCmd = values.getOrDefault("ciphertextCmd", null);
        this.ivCmd = values.getOrDefault("ivCmd", null);
        this.gcmAuthenticationTagCmd = values.getOrDefault("gcmAuthenticationTagCmd", null);
        this.plaintextData = values.getOrDefault("plaintextData", null);
        this.associatedPlaintextData = values.getOrDefault("associatedPlaintextData", null);
        this.ciphertextData = values.getOrDefault("ciphertextData", null);
        this.ivData = values.getOrDefault("ivData", null);
        this.gcmAuthenticationTagData = values.getOrDefault("gcmAuthenticationTagData", null);
    }

    public String getCurrentUTicketId() {
        return currentUTicketId;
    }

    public void setCurrentUTicketId(String currentUTicketId) {
        this.currentUTicketId = currentUTicketId;
    }

    public String getCurrentDeviceId() {
        return currentDeviceId;
    }

    public void setCurrentDeviceId(String currentDeviceId) {
        this.currentDeviceId = currentDeviceId;
    }

    public String getCurrentHolderId() {
        return currentHolderId;
    }

    public void setCurrentHolderId(String currentHolderId) {
        this.currentHolderId = currentHolderId;
    }

    public String getCurrentTaskScope() {
        return currentTaskScope;
    }

    public void setCurrentTaskScope(String currentTaskScope) {
        this.currentTaskScope = currentTaskScope;
    }

    public String getChallenge1() {
        return challenge1;
    }

    public void setChallenge1(String challenge1) {
        this.challenge1 = challenge1;
    }

    public String getChallenge2() {
        return challenge2;
    }

    public void setChallenge2(String challenge2) {
        this.challenge2 = challenge2;
    }

    public String getKeyExchangeSalt1() {
        return keyExchangeSalt1;
    }

    public void setKeyExchangeSalt1(String keyExchangeSalt1) {
        this.keyExchangeSalt1 = keyExchangeSalt1;
    }

    public String getKeyExchangeSalt2() {
        return keyExchangeSalt2;
    }

    public void setKeyExchangeSalt2(String keyExchangeSalt2) {
        this.keyExchangeSalt2 = keyExchangeSalt2;
    }

    public String getCurrentSessionKeyStr() {
        return currentSessionKeyStr;
    }

    public void setCurrentSessionKeyStr(String currentSessionKeyStr) {
        this.currentSessionKeyStr = currentSessionKeyStr;
    }

    public String getPlaintextCmd() {
        return plaintextCmd;
    }

    public void setPlaintextCmd(String plaintextCmd) {
        this.plaintextCmd = plaintextCmd;
    }

    public String getAssociatedPlaintextCmd() {
        return associatedPlaintextCmd;
    }

    public void setAssociatedPlaintextCmd(String associatedPlaintextCmd) {
        this.associatedPlaintextCmd = associatedPlaintextCmd;
    }

    public String getIvCmd() {
        return ivCmd;
    }

    public void setIvCmd(String ivCmd) {
        this.ivCmd = ivCmd;
    }

    public String getCiphertextCmd() {
        return ciphertextCmd;
    }

    public void setCiphertextCmd(String ciphertextCmd) {
        this.ciphertextCmd = ciphertextCmd;
    }

    public String getGcmAuthenticationTagCmd() {
        return gcmAuthenticationTagCmd;
    }

    public void setGcmAuthenticationTagCmd(String gcmAuthenticationTagCmd) {
        this.gcmAuthenticationTagCmd = gcmAuthenticationTagCmd;
    }

    public String getPlaintextData() {
        return plaintextData;
    }

    public void setPlaintextData(String plaintextData) {
        this.plaintextData = plaintextData;
    }

    public String getAssociatedPlaintextData() {
        return associatedPlaintextData;
    }

    public void setAssociatedPlaintextData(String associatedPlaintextData) {
        this.associatedPlaintextData = associatedPlaintextData;
    }

    public String getIvData() {
        return ivData;
    }

    public void setIvData(String ivData) {
        this.ivData = ivData;
    }

    public String getCiphertextData() {
        return ciphertextData;
    }

    public void setCiphertextData(String ciphertextData) {
        this.ciphertextData = ciphertextData;
    }

    public String getGcmAuthenticationTagData() {
        return gcmAuthenticationTagData;
    }

    public void setGcmAuthenticationTagData(String gcmAuthenticationTagData) {
        this.gcmAuthenticationTagData = gcmAuthenticationTagData;
    }

    public static String currentSessionToJsonStr(CurrentSession currentSession) {
        return gson.toJson(currentSession);
    }

    public static CurrentSession jsonStrToCurrentSession(String json) {
        try {
            return gson.fromJson(json, CurrentSession.class);
        } catch (Exception e) {
            String failureMsg = "NOT VALID JSON or VALID R TICKET SCHEMA";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw new RuntimeException(failureMsg);
        }
    }
}
